### ISSUE TITLE
`with` expressions regressed

### ISSUE THREAD

**Issue by maxime-didier** [dhall-lang/dhall-haskell#2597]
With dhall 1.41.2, it used to be possible to do the following:

```dhall
(Some { x = 1 }) with ?.x = 2
=> Some { x = 2 }

(Some { x = 1 }) with ?.x = "hello"
=> Some { x = "hello" }
```

With dhall 1.42.1, a restriction was added forbidding a `with` expression from changing the type of an optional (it is a breaking change that was not announced in the changelog). This restriction is buggy and both the above expressions fail to evaluate with:

```
Error: ❰with❱ cannot change the type of an ❰Optional❱ value
```

**Comment by winitzki** [comment]
Interesting. Let me check my Scala implementation (it was passing all standard tests).

**Comment by winitzki** [comment]
Yes, my Scala implementation fails both expressions. The error appears to be already in the Dhall standard in the typing judgments for this expression. Let me get the details.

**Comment by maxime-didier** [comment]
Interesting. Why the restriction ? Why wasn't a rule like the following adopted ?

```
Γ ⊢ e : Optional T₀
Γ, x : T₀ ⊢ x with ks… : T₁
────────────────────────────────── x ∉ FV(e)
Γ ⊢ e with ?.ks… = v : Optional T₁
```

**Comment by winitzki** [comment]
The intent was to prevent changing the types of values inside `Optional`. 

I don't really know the motivation; why is it allowed to do `{ x = { y = 1 } } with x.y = "abc"` but not allowed to do a similar replacement under `Optional`.

But that was the change. A `with` expression should not change the type, it may only change a value.


```dhall
Some { x = { y = 1 }} with ?.x.y = 2   -- Evaluates to Some { x = { y = 2 } }
Some { x = { y = 1 }} with ?.x = 0   -- Type error, it is not allowed to replace { y : Natural} with another type
```

That's how I understand the intention behind this change.

If this is correct, then the current rule is wrong. It should have been something like this:

```
Γ ⊢ e : Optional T₀
Γ, x : T₀ ⊢ x with ks… = v : T₁
T₀ ≡ T₁
────────────────────────────────── x ∉ FV(e)
Γ ⊢ e with ?.ks… = v : Optional T₀
```

**Comment by kukimik** [comment]
I'm just linking the PR that introduced this for reference: https://github.com/dhall-lang/dhall-lang/pull/1254.

**Comment by winitzki** [comment]
I have looked through the discussions in there, and I don't see any real motivation for prohibiting the type change. The use case was that you need access to record fields under `Some`. So, it seems to me that we should allow `(Some { x = 1 }) with ?.x = "hello"`, that will change the type from `Optional { x : Natural }` to `Optional { x : Text }`.

The usual `with` operation with records already can change the type.

Is there a reason to allow `(Some { x = 1 }) with ?.x = 2` (evaluating to `Some { x = 2 }`) but not allowing `(Some { x = 1 }) with ?.x = "hello"` (evaluating to `Some { x = "hello" }`?

**Comment by MonoidMusician** [comment]
what type `T` would you pass to `None T` during evaluation when it is not `Some v`? figuring out the right type would require typechecking during evaluation.

**Comment by winitzki** [comment]
> what type `T` would you pass to `None T` during evaluation when it is not `Some v`? figuring out the right type would require typechecking during evaluation.

Good point. Maybe that was the motivation for prohibiting the type change `Optional T`.

**Comment by Gabriella439** [comment]
Yeah, the reason the type is not supposed to change is so that evaluation doesn't need to infer the type.  However, it is still weird that `(Some { x = 1 }) with ?.x = 2` doesn't work.  That seems like a specification bug, probably.  In fact, [the original PR that added support for this](https://github.com/dhall-lang/dhall-lang/pull/1254) specifically cited that as one of the examples they were trying to support as valid Dhall code:

> ```dhall
> ⊢ (Some { x = 1 }) with ?.x = 2
>
> Some { x = 2 }
> ```

**Comment by winitzki** [comment]
@Gabriella439 Yes, I believe it is a specification bug that will produce the wrong behavior if implemented literally according to specification. I described that in a previous comment in this PR. I already fixed this in my Scala implementation.

**Comment by winitzki** [comment]
Could someone please help me setting up vscode with haskell plugin? I'm getting errors with cabal when I open the project.

```
I had trouble figuring out the exact GHC version for the project. Falling back to using 'ghc'.

Failed to find the GHC version of this Cabal project. Error when calling cabal exec -v0 -- ghc --print-libdir
```

I have a Mac M1 and everything installed (ghc, ghcup, cabal, stack). I can run `stack build` successfully.

**Comment by winitzki** [comment]
I'm working on fixing this regression. I discovered the following sample code that shows clearly why this is clearly a bug:

```dhall
Welcome to the Dhall v1.42.2 REPL! Type :help for more information.
⊢ (Some { a = 1 }) with ? = { a = 2 }

Some { a = 2 }

⊢ (Some { a = 1 }) with ?.a = { a = 2 }

Some { a.a = 2 }

⊢ (Some { a = 1 }) with ?.x.y = { a = 2 }

Some { a = 1, x.y.a = 2 }

⊢ (Some 1) with ?.a.b.c = 123

Some (1 with a.b.c = 123)

⊢ :let x =  (Some 1) with ?.a.b.c = 123

x : Optional Natural
```

The existing typechecking rule allows the expression `e with ?.k = v` when `e` has type `Optional T` and when `v` has the same type `T`; regardless of what `k` is! But the evaluation proceeds by replacing the value under `Some` regardless of what the actual types are. The last three examples show that Dhall does not actually prevent the type from being changed, and do not prevent nonsensical terms like `1 with a.b.c = 123` which are nevertheless typed as `Optional Natural`.

**Comment by winitzki** [comment]
PR made for the dhall lang standard, with tests https://github.com/dhall-lang/dhall-lang/pull/1400

**Comment by winitzki** [comment]
PR made for dhall-haskell to fix this regression. New tests pass. https://github.com/dhall-lang/dhall-haskell/pull/2650

**Comment by mmhat** [comment]
Fixed in #2650
